File: fabric/src/datagen/java/dev/mayaqq/estrogen/datagen/impl/recipes/minecraft/EstrogenCraftingRecipes.java
package dev.mayaqq.estrogen.datagen.impl.recipes.minecraft;

import com.simibubi.create.AllBlocks;
import com.simibubi.create.AllItems;
import com.simibubi.create.foundation.recipe.IRecipeTypeInfo;
import dev.mayaqq.estrogen.datagen.base.recipes.BaseRecipeProvider;
import dev.mayaqq.estrogen.datagen.base.platform.recipes.PlatformRecipeHelper;
import dev.mayaqq.estrogen.registry.EstrogenBlocks;
import dev.mayaqq.estrogen.registry.EstrogenItems;
import dev.mayaqq.estrogen.registry.EstrogenRecipes;
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput;
import net.minecraft.data.recipes.RecipeCategory;
import net.minecraft.data.recipes.ShapedRecipeBuilder;
import net.minecraft.data.recipes.ShapelessRecipeBuilder;
import net.minecraft.data.recipes.SpecialRecipeBuilder;
import net.minecraft.tags.ItemTags;

public class EstrogenCraftingRecipes extends BaseRecipeProvider {

    public EstrogenCraftingRecipes(FabricDataOutput output, PlatformRecipeHelper helper) {
        super(output, helper);
    }

    @Override
    protected void init() {
        add(ShapelessRecipeBuilder.shapeless(RecipeCategory.FOOD, EstrogenItems.ESTROGEN_CHIP_COOKIE)
                .requires(EstrogenItems.ESTROGEN_PILL)
                .unlockedBy(getHasName(EstrogenItems.ESTROGEN_PILL), has(EstrogenItems.ESTROGEN_PILL))
                .requires(AllItems.WHEAT_FLOUR)
                .requires(AllItems.BAR_OF_CHOCOLATE)
        );

        add(ShapelessRecipeBuilder.shapeless(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.ESTROGEN_PILL_BLOCK)
                .requires(EstrogenItems.ESTROGEN_PILL, 9)
                .unlockedBy(getHasName(EstrogenItems.ESTROGEN_PILL), has(EstrogenItems.ESTROGEN_PILL))
        );

        add(ShapelessRecipeBuilder.shapeless(RecipeCategory.FOOD, EstrogenItems.ESTROGEN_PILL, 9)
                .requires(EstrogenBlocks.ESTROGEN_PILL_BLOCK)
                .unlockedBy(getHasName(EstrogenBlocks.ESTROGEN_PILL_BLOCK), has(EstrogenBlocks.ESTROGEN_PILL_BLOCK))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.REDSTONE, EstrogenBlocks.CENTRIFUGE, 1)
                .define('P', AllBlocks.MECHANICAL_PUMP)
                .define('T', AllBlocks.FLUID_TANK)
                .define('C', getCommonTag("copper_plates"))
                .define('M', AllItems.PRECISION_MECHANISM)
                .pattern("CTC")
                .pattern("PMP")
                .pattern("CTC")
                .unlockedBy(getHasName(AllItems.PRECISION_MECHANISM), has(AllItems.PRECISION_MECHANISM))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.COOKIE_JAR, 1)
                .define('G', getCommonTag("glass_panes"))
                .define('Z', getCommonTag("zinc_nuggets"))
                .pattern("GZG")
                .pattern("G G")
                .pattern("GGG")
                .unlockedBy(getHasName(AllItems.ZINC_NUGGET), has(AllItems.ZINC_NUGGET))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.MISC, EstrogenItems.THIGH_HIGHS, 1)
                .define('F', EstrogenItems.MOTH_FUZZ)
                .pattern("FFF")
                .pattern("F F")
                .pattern("F F")
                .unlockedBy(getHasName(EstrogenItems.MOTH_FUZZ), has(EstrogenItems.MOTH_FUZZ))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.MOTH_WOOL, 1)
                .define('F', EstrogenItems.MOTH_FUZZ)
                .pattern("FF")
                .pattern("FF")
                .unlockedBy(getHasName(EstrogenItems.MOTH_FUZZ), has(EstrogenItems.MOTH_FUZZ))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.QUILTED_MOTH_WOOL, 4)
                .define('F', EstrogenBlocks.MOTH_WOOL)
                .pattern("FF")
                .pattern("FF")
                .unlockedBy(getHasName(EstrogenBlocks.MOTH_WOOL), has(EstrogenBlocks.MOTH_WOOL))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.MOTH_WOOL_CARPET, 3)
                .define('F', EstrogenBlocks.MOTH_WOOL)
                .pattern("FF")
                .unlockedBy(getHasName(EstrogenBlocks.MOTH_WOOL), has(EstrogenBlocks.MOTH_WOOL))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.QUILTED_MOTH_WOOL_CARPET, 3)
                .define('F', EstrogenBlocks.QUILTED_MOTH_WOOL)
                .pattern("FF")
                .unlockedBy(getHasName(EstrogenBlocks.QUILTED_MOTH_WOOL), has(EstrogenBlocks.QUILTED_MOTH_WOOL))
        );

        add(ShapelessRecipeBuilder.shapeless(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.MOTH_SEAT)
                .requires(EstrogenBlocks.QUILTED_MOTH_WOOL, 1)
                .requires(ItemTags.WOODEN_SLABS)
                .unlockedBy(getHasName(EstrogenBlocks.MOTH_WOOL), has(EstrogenBlocks.MOTH_WOOL)
        ));

        add(ShapelessRecipeBuilder.shapeless(RecipeCategory.MISC, EstrogenItems.MOTH_FUZZ, 4)
                .requires(EstrogenBlocks.MOTH_WOOL)
                .unlockedBy(getHasName(EstrogenBlocks.MOTH_WOOL), has(EstrogenBlocks.MOTH_WOOL)
        ));

        add(ShapedRecipeBuilder.shaped(RecipeCategory.DECORATIONS, EstrogenBlocks.MOTH_BED, 1)
                .define('M', EstrogenBlocks.MOTH_WOOL)
                .define('W', ItemTags.PLANKS)
                .pattern("MMM")
                .pattern("WWW")
                .unlockedBy(getHasName(EstrogenBlocks.MOTH_WOOL), has(EstrogenBlocks.MOTH_WOOL))
        );

        add(ShapedRecipeBuilder.shaped(RecipeCategory.DECORATIONS, EstrogenBlocks.QUILTED_MOTH_BED, 1)
                .define('M', EstrogenBlocks.QUILTED_MOTH_WOOL)
                .define('W', ItemTags.PLANKS)
                .pattern("MMM")
                .pattern("WWW")
                .unlockedBy(getHasName(EstrogenBlocks.QUILTED_MOTH_WOOL), has(EstrogenBlocks.QUILTED_MOTH_WOOL))
        );

        add(exporter -> SpecialRecipeBuilder.special(EstrogenRecipes.THIGH_HIGH_DYE_SERIALIZER.get())
                .save(exporter, "estrogen:thigh_high_dye")
        );
    }

    @Override
    protected IRecipeTypeInfo getRecipeType() {
        return null;
    }

    @Override
    public String getRecipeName() {
        return "crafting";
    }
}

File: src/fabric/data/kotlin/dev/mayaqq/estrogen/datagen/impl/recipes/minecraft/EstrogenCraftingRecipes.kt
package dev.mayaqq.estrogen.datagen.impl.recipes.minecraft

import dev.mayaqq.estrogen.content.EstrogenBlocks
import dev.mayaqq.estrogen.content.EstrogenItems
import dev.mayaqq.estrogen.content.recipes.DreamCatcherDyeRecipe
import dev.mayaqq.estrogen.content.recipes.ThighHighDyeRecipe
import dev.mayaqq.estrogen.datagen.api.platform.PlatformRecipeHelper
import dev.mayaqq.estrogen.id
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput
import net.fabricmc.fabric.api.datagen.v1.provider.FabricRecipeProvider
import net.minecraft.advancements.CriteriaTriggers
import net.minecraft.advancements.critereon.ImpossibleTrigger
import net.minecraft.core.HolderLookup
import net.minecraft.data.recipes.*
import net.minecraft.tags.ItemTags
import net.minecraft.world.item.Items
import java.util.concurrent.CompletableFuture

class EstrogenCraftingRecipes(output: FabricDataOutput, lookup: CompletableFuture<HolderLookup.Provider>, val helper: PlatformRecipeHelper) : FabricRecipeProvider(output, lookup) {
    override fun buildRecipes(output: RecipeOutput) {
        ShapedRecipeBuilder.shaped(RecipeCategory.MISC, EstrogenBlocks.ColonThreeBlock.value!!)
            .define('C', EstrogenItems.ColonThree)
            .define('E', Items.EGG)
            .pattern("CCC")
            .pattern("CEC")
            .pattern("CCC")
            .showNotification(false)
            .unlockedBy("never", CriteriaTriggers.IMPOSSIBLE.createCriterion(ImpossibleTrigger.TriggerInstance()))
            .save(output, id("colon_three_manual_only"))
        ShapelessRecipeBuilder.shapeless(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.EstrogenPillBlock.value!!)
            .requires(EstrogenItems.EstrogenPill, 9)
            .unlockedBy(getHasName(EstrogenItems.EstrogenPill), has(EstrogenItems.EstrogenPill))
            .save(output)
        ShapelessRecipeBuilder.shapeless(RecipeCategory.FOOD, EstrogenItems.EstrogenPill, 9)
            .requires(EstrogenBlocks.EstrogenPillBlock.value!!)
            .unlockedBy(getHasName(EstrogenBlocks.EstrogenPillBlock.value!!), has(EstrogenBlocks.EstrogenPillBlock.value!!))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.CookieJar.value!!, 1)
            .define('G', helper.commonTag("glass_panes"))
            .define('Z', helper.commonTag("nuggets/iron"))
            .pattern("GZG")
            .pattern("G G")
            .pattern("GGG")
            .unlockedBy(getHasName(EstrogenItems.EstrogenChipCookie), has(EstrogenItems.EstrogenChipCookie))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.MISC, EstrogenItems.ThighHighs, 1)
            .define('F', EstrogenItems.MothFuzz)
            .pattern("FFF")
            .pattern("F F")
            .pattern("F F")
            .unlockedBy(getHasName(EstrogenItems.MothFuzz), has(EstrogenItems.MothFuzz))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.MothWool.value!!, 1)
            .define('F', EstrogenItems.MothFuzz)
            .pattern("FF")
            .pattern("FF")
            .unlockedBy(getHasName(EstrogenItems.MothFuzz), has(EstrogenItems.MothFuzz))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.QuiltedMothWool.value!!, 4)
            .define('F', EstrogenBlocks.MothWool.value!!)
            .pattern("FF")
            .pattern("FF")
            .unlockedBy(getHasName(EstrogenBlocks.MothWool.value!!), has(EstrogenBlocks.MothWool.value!!))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.MothCarpet.value!!, 3)
            .define('F', EstrogenBlocks.MothWool.value!!)
            .pattern("FF")
            .unlockedBy(getHasName(EstrogenBlocks.MothWool.value!!), has(EstrogenBlocks.MothWool.value!!))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.BUILDING_BLOCKS, EstrogenBlocks.QuiltedMothCarpet.value!!, 3)
            .define('F', EstrogenBlocks.QuiltedMothWool.value!!)
            .pattern("FF")
            .unlockedBy(getHasName(EstrogenBlocks.QuiltedMothWool.value!!), has(EstrogenBlocks.QuiltedMothWool.value!!))
            .save(output)
        ShapelessRecipeBuilder.shapeless(RecipeCategory.MISC, EstrogenItems.MothFuzz, 4)
            .requires(EstrogenBlocks.MothWool.value!!)
            .unlockedBy(getHasName(EstrogenBlocks.MothWool.value!!), has(EstrogenBlocks.MothWool.value!!))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.DECORATIONS, EstrogenBlocks.MothBed.value!!, 1)
            .define('M', EstrogenBlocks.MothWool.value!!)
            .define('W', ItemTags.PLANKS)
            .pattern("MMM")
            .pattern("WWW")
            .unlockedBy(getHasName(EstrogenBlocks.MothWool.value!!), has(EstrogenBlocks.MothWool.value!!))
            .save(output)
        ShapedRecipeBuilder.shaped(RecipeCategory.DECORATIONS, EstrogenBlocks.QuiltedMothBed.value!!, 1)
            .define('M', EstrogenBlocks.QuiltedMothWool.value!!)
            .define('W', ItemTags.PLANKS)
            .pattern("MMM")
            .pattern("WWW")
            .unlockedBy(getHasName(EstrogenBlocks.QuiltedMothWool.value!!), has(EstrogenBlocks.QuiltedMothWool.value!!))
            .save(output)
        SpecialRecipeBuilder.special(::ThighHighDyeRecipe)
            .save(output, "estrogen:thigh_high_dye")
        SpecialRecipeBuilder.special(::DreamCatcherDyeRecipe)
            .save(output, "estrogen:dreamcatcher_dye")
        ShapedRecipeBuilder.shaped(RecipeCategory.DECORATIONS, EstrogenBlocks.DreamCatcher.value!!, 1)
            .define('S', Items.STRING)
            .define('C', Items.COBWEB)
            .define('W', Items.STICK)
            .define('F', Items.FEATHER)
            .pattern("WSW")
            .pattern("WCW")
            .pattern("FFF")
            .unlockedBy(getHasName(Items.COBWEB), has(Items.COBWEB))
            .save(output)
    }
}